File: snippets/cpp/cycle/cycleundir.cpp
vi find_cycle_undir(vector<vi> &g, int s) {
    int n = sza(g); vi vis(n), par(n, -1); vi cycle;
    function<bool(int, int)> dfs = [&](int u, int p) {
        vis[u] = 1;
        for (int v : g[u]) {
            if (v == p) continue;
            if (vis[v]) {
                for (int x = u; x != v; x = par[x]) cycle.pb(x);
                cycle.pb(v); reverse(all(cycle));
                return true;
            }
            par[v] = u;
            if (dfs(v, u)) return true;
        }
        return false;
    };
    for (int i = 0; i < n; i++)
        if (!vis[i] && dfs(i, -1)) return cycle;
    return {};
}

File: snippets/rust/cycle/cycleundir.rs
// Find a cycle in an undirected graph. Returns nodes of the cycle, or empty.
// Edges in `g` should not duplicate (use one entry per edge in each direction).
fn find_cycle_undir(g: &Vec<Vec<usize>>) -> Vec<usize> {
    let n = g.len();
    let mut vis = vec![false; n];
    let mut par = vec![usize::MAX; n];
    let mut stk: Vec<(usize, usize, usize)> = Vec::new();
    for start in 0..n {
        if vis[start] { continue; }
        vis[start] = true;
        stk.push((start, usize::MAX, 0));
        while let Some(&(u, p, i)) = stk.last() {
            if i < g[u].len() {
                stk.last_mut().unwrap().2 += 1;
                let v = g[u][i];
                if v == p { continue; }
                if vis[v] {
                    let mut cycle = vec![v];
                    let mut x = u;
                    while x != v { cycle.push(x); x = par[x]; }
                    cycle.reverse();
                    return cycle;
                }
                vis[v] = true;
                par[v] = u;
                stk.push((v, u, 0));
            } else {
                stk.pop();
            }
        }
    }
    Vec::new()
}
